### PR TITLE
KLayout Step Unit Tests

### LINKED ISSUE #234: Multi-threaded XOR not functional
### Description

`KLayout.XOR` always runs on a maximum of one thread when inspected via `psutils` or `top`.

This is due to two things:

1. I forgor to pass the option to the command-line
2. https://github.com/KLayout/klayout/issues/1517

### Expected Behavior

KLayout XOR is an expensive process and should run on multiple threads

### Environment report

```sh
kernel: Linux
kernel_version: 6.1.52
supported: True
distro: nixos
distro_version: 23.05
python_version: 3.10.9
python_path:
  - /nix/store/savf1qa5abvpvwhvz5fch7bggr3n7br0-python3.10-openlane/bin
  - /nix/store/ns7p6q9gw667v41ysn750103v7dgxgwz-python3.10-click-8.1.3/lib/python3.10/site-packages
  - /nix/store/n0kq2qalwygww8h3pr7v6av7h7jxgd25-python3.10-cloup-1.0.1/lib/python3.10/site-packages
  - /nix/store/l8y1hkbki6pjbr1lm0k0iv51ii9864wz-python3.10-pyyaml-6.0/lib/python3.10/site-packages
  - /nix/store/gkphk53397gfy6mr5nhlqyg5cmgk304h-python3.10-rich-12.5.1/lib/python3.10/site-packages
  - /nix/store/i09vhwzivdhr017688ywhzjwz14xdhn9-python3.10-requests-2.28.1/lib/python3.10/site-packages
  - /nix/store/yrp4f59aa0a1kvrgzdm58v56q8h90604-python3.10-pcpp-1.21/lib/python3.10/site-packages
  - /nix/store/25dlz6av123fbc6qclh6w2agq62dlkmh-python3.10-volare/lib/python3.10/site-packages
  - /nix/store/qvb8isbg9kgmjzkmm9i20r05hns807cj-python3.10-tkinter-3.10.9/lib/python3.10/site-packages
  - /nix/store/xkvl0krzk8l9ch8ksxkb9xmg2jjiszq7-python3.10-lxml-4.9.1/lib/python3.10/site-packages
  - /nix/store/cdgsnj5kmpawp52iramjc7craindq2hr-python3.10-deprecated-1.2.13/lib/python3.10/site-packages
  - /nix/store/aip2pp28ld9wcbqgwsgy3xybqhc21025-python3.10-immutabledict-2.2.2/lib/python3.10/site-packages
  - /nix/store/qll21g8p4421c0bw98qs8rg9gr64256g-python3.10-libparse/lib/python3.10/site-packages
  - /nix/store/c2ahwjcjnhi2r9yw42fa7l66gnvqsbp7-python3.10-psutil-5.9.3/lib/python3.10/site-packages
  - /nix/store/vxl992skg2q9s5wxpjsg972i0556ngxq-yosys-0.23/lib/python3.10/site-packages
  - /nix/store/96165i71r0l69b1n3pn8z4qksby5g30c-openroad/lib/python3.10/site-packages
  - /nix/store/yizhphjr22q1rr77n1067vyykq520gdd-klayout-6a36bfa7c04f55bd732f8e0f91b553c8f9cebed7/lib/python3.10/site-packages
  - /nix/store/y3ymwmqpmarqqqxp537jsgiqf1mp5rqz-netgen/lib/python3.10/site-packages
  - /nix/store/1xfp6kmcah4pmic46v2s2nbgmm08ncv0-magic-vlsi/lib/python3.10/site-packages
  - /nix/store/z41wwdc6zlmxz0mgcjmg33gmdcxwsww4-iverilog-11.0/lib/python3.10/site-packages
  - /nix/store/z9sqf8a4vy55416cbxz40wz6mdhjw946-verilator/lib/python3.10/site-packages
  - /nix/store/378rxqdx6lh4swhw49ifbilc6hbk7h2r-tcl-8.6.11/lib/python3.10/site-packages
  - /nix/store/kcqiwya701lhcwhl1gci4cmjk7c5bcav-surelog-1.76/lib/python3.10/site-packages
  - /nix/store/pjjdz5sjrf2xk8vzf7461wbhkky00k4r-python3.10-cffi-1.15.1/lib/python3.10/site-packages
  - /nix/store/6i9gf8hz2h9as3lxsraysv7qjh102g67-python3.10-pycparser-2.21/lib/python3.10/site-packages
  - /nix/store/65cp4izx3bllnwqn7c7dhrq9h9gmjkal-python3-3.10.9/lib/python3.10/site-packages
  - /nix/store/3gb94vhj1ycgvn135h19qh5zlr39mr2f-python3.10-argon2-cffi-bindings-21.2.0/lib/python3.10/site-packages
  - /nix/store/qhbhagpm0wz32r8gzddg6ykcsc1lsy51-python3-3.10.9-env/lib/python3.10/site-packages
  - /nix/store/22pfzrq241ipn89330825ybdildl23wx-python3.10-notebook-6.4.12/lib/python3.10/site-packages
  - /nix/store/19n7yd6drwv783w65yf7h7mcrfaagpb7-python3.10-Jinja2-3.1.2/lib/python3.10/site-packages
  - /nix/store/j0icmszh4azcmikmrzib4vicnwns6fn9-python3.10-babel-2.11.0/lib/python3.10/site-packages
  - /nix/store/yfxxhqmckk7kljaqydz07xjfhv3wnx2d-python3.10-pytz-2022.2.1/lib/python3.10/site-packages
  - /nix/store/yfis2i220a8pgy53kx0g5wqwdl7hjn42-python3.10-markupsafe-2.1.1/lib/python3.10/site-packages
  - /nix/store/5jphlf51x69jjlj4did10nd9f68xnml0-python3.10-tornado-6.2/lib/python3.10/site-packages
  - /nix/store/gmac070bibirfdf5wysk4ch5r2hvfnng-python3.10-ipython_genutils-0.2.0/lib/python3.10/site-packages
  - /nix/store/g2yv0ykq3baqrfza664rivffd96sicrw-python3.10-traitlets-5.4.0/lib/python3.10/site-packages
  - /nix/store/v1l9pwi92fbnjvsd0mpck2xd1cjigvsc-python3.10-decorator-5.1.1/lib/python3.10/site-packages
  - /nix/store/kx5anc33hbjj9cd9h3wll51p9b6lasb4-python3.10-six-1.16.0/lib/python3.10/site-packages
  - /nix/store/bfh8fgpcpskjiss1diq92yxvfhw0mvhg-python3.10-jupyter_core-4.11.2/lib/python3.10/site-packages
  - /nix/store/20ckyqfx5hf6riilgzv0sr26jva9ws9i-python3.10-Send2Trash-1.8.1b0/lib/python3.10/site-packages
  - /nix/store/big6ccqb03w6ilr819ra9n9mhznv1j6z-python3.10-jupyter_client-7.3.5/lib/python3.10/site-packages
  - /nix/store/9q0sjf2f0dpkwdx14f755fhqnx81lsiv-python3.10-entrypoints-0.4/lib/python3.10/site-packages
  - /nix/store/njq5zmps7r5shv9py91agng07li32rv3-python3.10-nest_asyncio-1.5.5/lib/python3.10/site-packages
  - /nix/store/rfq9y7g59q4lwm4r1mw7b4p7pvlr84xw-python3.10-python-dateutil-2.8.2/lib/python3.10/site-packages
  - /nix/store/l0rky4yga6ji2pk44mvvv34gdz3i270f-python3.10-pyzmq-23.2.1/lib/python3.10/site-packages
  - /nix/store/x6vjsnm4aq0w1h4l9mi5pn7wq26p020k-python3.10-py-1.11.0/lib/python3.10/site-packages
  - /nix/store/ifi68lsz88dzaiy4dar51z0nyai3kk6z-python3.10-nbformat-5.7.0/lib/python3.10/site-packages
  - /nix/store/clp6zb4s5h8wicamcxkkrwq3091kknjj-python3.10-fastjsonschema-2.16.2/lib/python3.10/site-packages
  - /nix/store/vghm2kpw39j1mx8l38gsamy4ricqpdwf-python3.10-jsonschema-4.17.0/lib/python3.10/site-packages
  - /nix/store/h7j7mrqva6wfsc9a3c4sbgs6dh0zd5m0-python3.10-attrs-22.1.0/lib/python3.10/site-packages
  - /nix/store/ybziay9fz0cy1nx3bvqi1rjcg9fn1gyz-python3.10-pyrsistent-0.18.1/lib/python3.10/site-packages
  - /nix/store/s1nwcpdnpgp294iivj7pp5hvlwvf8lc5-python3.10-nbconvert-7.2.3/lib/python3.10/site-packages
  - /nix/store/vsfv1csakh0hvphcxk3hdk8hlm8vm9f4-python3.10-beautifulsoup4-4.11.1/lib/python3.10/site-packages
  - /nix/store/mfdissddhv02i6ly0wi1s5cv8r6i5kyd-python3.10-chardet-5.0.0/lib/python3.10/site-packages
  - /nix/store/wk1zvimzsc89rbr6nqn3g5bjnq5ggjcv-python3.10-html5lib-1.1/lib/python3.10/site-packages
  - /nix/store/risl3q040vj1c6gl4xgkdbvy7n94dr3r-python3.10-webencodings-0.5.1/lib/python3.10/site-packages
  - /nix/store/3rjrab93xa8y23yd1m2aaiydd78342wv-python3.10-soupsieve-2.3.2.post1/lib/python3.10/site-packages
  - /nix/store/raw0xqlbzq2vlsjh2n0sd8775fnfrjyb-python3.10-bleach-5.0.1/lib/python3.10/site-packages
  - /nix/store/myyr3gvkqc7qdcwz5rgp63as3gchcsw5-python3.10-packaging-21.3/lib/python3.10/site-packages
  - /nix/store/3nwm7x7vsc4df6qwzr39wg94p23rzqgl-python3.10-pyparsing-3.0.9/lib/python3.10/site-packages
  - /nix/store/49pg4ljx6qandwdr6mbc8w8k8fh0dd2j-python3.10-setuptools-65.3.0/lib/python3.10/site-packages
  - /nix/store/b6ld8s8mj11sisxv1si2mdq7wy6l9rz5-python3.10-defusedxml-0.7.1/lib/python3.10/site-packages
  - /nix/store/mfhym26ywy1iaqdlhx1m8irz92z8dh8z-python3.10-jupyterlab_pygments-0.1.2/lib/python3.10/site-packages
  - /nix/store/cpv31grqdbxvsq7jx4kdgbxg2rsxkkfy-python3.10-pygments-2.13.0/lib/python3.10/site-packages
  - /nix/store/9ljcxsc2gqnf5gysl45993v4v88iwzz7-python3.10-docutils-0.19/lib/python3.10/site-packages
  - /nix/store/v5fgbri0rdjw183c8z5cz04l4h3jahmc-python3.10-mistune-2.0.4/lib/python3.10/site-packages
  - /nix/store/gprmixpnzsyyvl1ab5w4j5x2gzqds0dz-python3.10-nbclient-0.6.8/lib/python3.10/site-packages
  - /nix/store/zwny3mk377qlpx88lpskaxkb1jy8dyaf-python3.10-async-generator-1.10/lib/python3.10/site-packages
  - /nix/store/w6pcnr7y354pkd7s36b5didyyfzgn1i9-python3.10-pandocfilters-1.5.0/lib/python3.10/site-packages
  - /nix/store/dlz3pszi4ly71s5ymijhkkzml9g2i1v0-python3.10-tinycss2-1.1.1/lib/python3.10/site-packages
  - /nix/store/z2dhmf1y3bm4pfrrf2pb4m4jl5s036kg-python3.10-ipykernel-6.15.3/lib/python3.10/site-packages
  - /nix/store/d4swckgazsa236wdixkvpyxz2gjb8knj-python3.10-ipython-8.4.0/lib/python3.10/site-packages
  - /nix/store/3q71i6346iwx4azyx23xd4337wbk5j39-python3.10-backcall-0.2.0/lib/python3.10/site-packages
  - /nix/store/350ybb2fxm4fbfwqi4zibmz9qs6yszbc-python3.10-jedi-0.18.1/lib/python3.10/site-packages
  - /nix/store/3bkfagkxrggcbkdxlyym8f4pfmdaml74-python3.10-parso-0.8.3/lib/python3.10/site-packages
  - /nix/store/fmjwj7m3cd2s5d57vbb7jjfk6sxjxm68-python3.10-matplotlib-inline-0.1.6/lib/python3.10/site-packages
  - /nix/store/h7cmdfyv11sg369lkglhz3lrg97wx0af-python3.10-pexpect-4.8.0/lib/python3.10/site-packages
  - /nix/store/bihdcprlyf9jkrnjpsgpp2i9pf2cqib1-python3.10-ptyprocess-0.7.0/lib/python3.10/site-packages
  - /nix/store/r51iq868vlvl9wnvwqjszzw4gp9wwlc6-python3.10-pickleshare-0.7.5/lib/python3.10/site-packages
  - /nix/store/04m2m7zd2axx021fbw9ix0lnhafwgq9k-python3.10-path-16.4.0/lib/python3.10/site-packages
  - /nix/store/dvx6vkz7mg9k6j80vrgd5bxywdfbqhzc-python3.10-prompt-toolkit-3.0.31/lib/python3.10/site-packages
  - /nix/store/4hz6djp5my9mj8x594s7mv693a8zy6sk-python3.10-wcwidth-0.2.5/lib/python3.10/site-packages
  - /nix/store/32kvwf0n7m63r654avn44c4vvzb073y0-python3.10-stack-data-0.2.0/lib/python3.10/site-packages
  - /nix/store/42gmwq7yfp8qam09daz5ribzisw1nh3a-python3.10-asttokens-2.0.8/lib/python3.10/site-packages
  - /nix/store/f04l1d9bm4wfb5fmfxcfll4qxihgnihs-python3.10-executing-0.8.2/lib/python3.10/site-packages
  - /nix/store/qinnsnlb8j423zy90mi23sssxgayr6z7-python3.10-pure_eval-0.2.2/lib/python3.10/site-packages
  - /nix/store/yqziisvama0a8vga5lxp70iya0znfbwv-python3.10-toml-0.10.2/lib/python3.10/site-packages
  - /nix/store/691d7j7ssm1j7ybhxfqzr52cpgl0i6q8-python3.10-terminado-0.15.0/lib/python3.10/site-packages
  - /nix/store/6bkqcc8d265j605rzrb92lr2pn82442h-python3.10-brotlicffi-1.0.9.2/lib/python3.10/site-packages
  - /nix/store/8k41q7l3ff3i4lz0drn34w3phk8fw49z-python3.10-certifi-2022.12.07/lib/python3.10/site-packages
  - /nix/store/2s5sa0zibgjshn0s3hhjna4cql3flsnn-python3.10-charset-normalizer-2.1.0/lib/python3.10/site-packages
  - /nix/store/wl8rwlabb29g5gqpy0dsqf9g4xagnc2j-python3.10-idna-3.4/lib/python3.10/site-packages
  - /nix/store/3whxw66aiiixk326lj6h63cxbv0cpilg-python3.10-urllib3-1.26.12/lib/python3.10/site-packages
  - /nix/store/vwrdsq0v60p5gv5v6vlydmy045ihf29q-python3.10-brotli-1.0.9/lib/python3.10/site-packages
  - /nix/store/ah2w40lqazbkjd4cga7binijihm8dvb3-python3.10-pysocks-1.7.1/lib/python3.10/site-packages
  - /nix/store/16riacyavyzhav5ly0f4spkwkj832yqd-python3.10-prometheus-client-0.15.0/lib/python3.10/site-packages
  - /nix/store/97zsmfrirvvikjqpm3lp9a8vfhnx1jsr-python3.10-argon2-cffi-21.3.0/lib/python3.10/site-packages
  - /nix/store/65cp4izx3bllnwqn7c7dhrq9h9gmjkal-python3-3.10.9/lib/python310.zip
  - /nix/store/65cp4izx3bllnwqn7c7dhrq9h9gmjkal-python3-3.10.9/lib/python3.10
  - /nix/store/65cp4izx3bllnwqn7c7dhrq9h9gmjkal-python3-3.10.9/lib/python3.10/lib-dynload
  - /nix/store/savf1qa5abvpvwhvz5fch7bggr3n7br0-python3.10-openlane/lib/python3.10/site-packages
  - /nix/store/k5b7jpzi9dvg3x7h2a90df14aq8wh6bl-python3.10-commonmark-0.9.1/lib/python3.10/site-packages
  - /nix/store/8zzhma74gr86jzakxd8f1pzlrdf8snzx-python3.10-future-0.18.2/lib/python3.10/site-packages
  - /nix/store/filc9bjiifd21l66dicrwr4y3rbza01x-python3.10-zstandard-0.19.0/lib/python3.10/site-packages
  - /nix/store/hjhzx3ic251y5gjigcxiyp2chdzd8s9k-python3.10-wrapt-1.13.3/lib/python3.10/site-packages
tkinter: True
container_info:
  engine: docker
  version: 20.10.25
  conmon: False
  rootless: True
nix_info:
  version: nix-env (Nix) 2.13.5
  multi_user: True
  sandbox: True
  channels:
    root:
      - nixos-23.05
    donn: []
  nixpkgs: /nix/var/nix/profiles/per-user/root/channels/nixos
```

### Reproduction material

N/A

### Relevant log output

```sh
N/A
```

### PR DESCRIPTION
**Pull request by donn**
* Created unit testing for all KLayout steps
* Fixed `KLayout.XOR` not properly supporting threading
  * `KLAYOUT_XOR_THREADS` is now optional, with the thread count being equal to your thread count by default
  * Added new variable, `KLAYOUT_XOR_TILE_SIZE`, which is the size of the side of a tile in microns (the tile size must be sufficiently smaller than the design for KLayout to bother threading)
  * Added `info` prints for thread kounts for both `KLayout.XOR` and `OpenROAD.DetailedRouting`
* Internal change to unit testing infrastructure, which copies `base.sdc` by default to all folders to avoid the file being in every single damn test
---

Depends on https://github.com/efabless/openlane2-step-unit-tests/pull/5
Resolves #234

### CODE REVIEW

File: openlane/steps/klayout.py [hunk 1]
```diff
@@ -285,6 +290,12 @@ def run(self, state_in: State, **kwargs) -> Tuple[ViewsUpdate, MetricsUpdate]:
 
         kwargs, env = self.extract_env(kwargs)
 
+        tile_size_options = []
+        if tile_size := self.config["KLAYOUT_XOR_TILE_SIZE"]:
+            tile_size_options += ["--tile-size", str(tile_size)]
+
+        thread_count = self.config["KLAYOUT_XOR_THREADS"] or os.cpu_count() or 1
+
```
Comment (kareefardi) [inline openlane/steps/klayout.py:298]: Can you add a warning here similar to https://github.com/efabless/openlane2/blob/fbfe20ebf95e1b3e9230821ea9c6d9b6f7ef9989/openlane/steps/openroad.py#L843

Comment (donn) [reply openlane/steps/klayout.py:298]: It's not really a warning *per se* (warnings are more for things that are unrecommended), but we can add an "info" call

Comment (kareefardi) [reply openlane/steps/klayout.py:298]: Sure. Variables or arguments set by OL which aren't saved with the step state should be printed at least.

File: openlane/scripts/klayout/xor.drc [hunk 2]
```diff
@@ -83,14 +99,14 @@ ignore_list = $ignore.split(";")
...
+layers.sort.uniq.each do |layer_name, layer_data|
   if ignore_list.include? layer_name
-    warn "Skipping #{layer_name}"
+    warn "--- Skipping #{layer_name}… ---"
   else
     info "--- Running XOR for layer #{layer_name} ---"
 
     layer_info = layers[layer_name]
-    xor_data = a.input(layer_name) ^ b.input(layer_name)
+    xor_data = (a.input(layer_name, layer_data) ^ b.input(layer_name, layer_data))
```
Comment (kareefardi) [inline openlane/scripts/klayout/xor.drc:109]: I don't think this is needed.

File: test/steps/test_all_steps.py [hunk 3]
```diff
@@ -100,6 +100,9 @@ def test_step_folder(test: str, pdk_root: str):
         os.path.join(os.getcwd(), "handler.py"), "handle"
     ) or (lambda: None)
 
+    base_sdc = os.path.join(get_script_dir(), "base.sdc")
```
Comment (kareefardi) [inline test/steps/test_all_steps.py:103]: A bit magical but OK.

File: openlane/scripts/klayout/xor.drc (hunk 2 above)
Comment (donn) [reply openlane/scripts/klayout/xor.drc:109]: I tried to match the KLayout XOR script written by Mathias a bit better.

File: test/steps/test_all_steps.py (hunk 3 above)
Comment (donn) [reply test/steps/test_all_steps.py:103]: Better than having the file 80 times. Also it's unit tests, who cares

**Review by kareefardi — APPROVED** [review]

**Review by kareefardi — APPROVED** [review]

**Pull request merged**